internal/server/error.go

```diff
@@ -0,0 +1,35 @@
+package server
+
+import (
+	"encoding/json"
+	"net/http"
+)
+
+type errJSONWrapper struct {
+	Error string `json:"error"`
+}
+
+func httpError(w http.ResponseWriter, status int, errString string) {
+	w.WriteHeader(status)
+	if errString == "" {
+		errString = http.StatusText(status)
+	}
+	body := errJSONWrapper{Error: errString}
+	_ = json.NewEncoder(w).Encode(body)
+}
+
+type errorsJSONWrapper struct {
+	Errors []string `json:"errors"`
+}
+
+func httpErrors(w http.ResponseWriter, status int, errors []error) {
+	w.WriteHeader(status)
+
+	errs := make([]string, len(errors))
```

Comment:
if you make() you need to close() somewhere, right?

Reply:
Ha that's a fair question! You don't need to close slices (=dynamic arrays in Go) in Go, they just get garbage collected.

You can however close *channels* (like `make(chan string)`) but that's more of a developer thing rather than a performance/memory thing. Someone explained it better than me though:

> You needn't close every channel when you've finished with it. It's only necessary to close a channel when it is important to tell the receiving goroutines that all data have been sent. A channel that the garbage collector determinies to be unreachable will have its resources reclaimed whether or not it is closed. (Don't confuse this with the close operation for open files. It is important to call the Close method on every file when you've finished with it.)

Reply:
thank you for taking the time to explain